wavefront/timeUniformSample.go

```diff
@@ -0,0 +1,227 @@
+	sample := &TimeUniformSample{lifetime: lifetime, values: newQueue(size)}
+
+	if ticker == nil {
+		ticker = time.NewTicker(recycleTime)
+		go func() {
+			for t := range ticker.C {
+				fmt.Println("Tick at", t)
+				for _, sample := range sampleList {
+					sample.cleanOldValues()
+				}
+			}
+		}()
+	}
+
+	sampleList = append(sampleList, sample)
+	return sample
+}
+
+// Clear clears all samples.
+func (s *TimeUniformSample) Clear() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.values = Queue{}
+}
+
+// Count returns the number of samples recorded, which may exceed the
+// reservoir size.
+func (s *TimeUniformSample) Count() int64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return int64(s.values.len)
+}
+
+// Max returns the maximum value in the sample, which may not be the maximum
+// value ever to be part of the sample.
+func (s *TimeUniformSample) Max() int64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SampleMax(s.rawValues())
+}
+
+// Mean returns the mean of the values in the sample.
+func (s *TimeUniformSample) Mean() float64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SampleMean(s.rawValues())
+}
+
+// Min returns the minimum value in the sample, which may not be the minimum
+// value ever to be part of the sample.
+func (s *TimeUniformSample) Min() int64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SampleMin(s.rawValues())
+}
+
+// Percentile returns an arbitrary percentile of values in the sample.
+func (s *TimeUniformSample) Percentile(p float64) float64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SamplePercentile(s.rawValues(), p)
+}
+
+// Percentiles returns a slice of arbitrary percentiles of values in the
+// sample.
+func (s *TimeUniformSample) Percentiles(ps []float64) []float64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SamplePercentiles(s.rawValues(), ps)
+}
+
+// Size returns the size of the sample, which is at most the reservoir size.
+func (s *TimeUniformSample) Size() int {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return len(s.rawValues())
+}
+
+// Snapshot returns a read-only copy of the sample.
+func (s *TimeUniformSample) Snapshot() metrics.Sample {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	values := make([]int64, s.values.len)
+	copy(values, s.rawValues())
+	return metrics.NewSampleSnapshot(int64(s.values.len), values)
+}
+
+// StdDev returns the standard deviation of the values in the sample.
+func (s *TimeUniformSample) StdDev() float64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SampleStdDev(s.rawValues())
+}
+
+// Sum returns the sum of the values in the sample.
+func (s *TimeUniformSample) Sum() int64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SampleSum(s.rawValues())
+}
+
+// Update samples a new value.
+func (s *TimeUniformSample) Update(v int64) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.cleanOldValues()
```

Comment:
Once you introduce locking in the `cleanOldValues` method, you'll have to move this call prior to locking within this function as Golang `Mutex` is not reentrant.